lib/common/async_io.c

```diff
@@ -0,0 +1,290 @@
+        if ((queue->head = popped->next) == NULL)
+            queue->tail = &queue->head;
+        --queue->size;
+        popped->next = NULL;
+    }
+    return popped;
+}
+
+static int submit_commands(h2o_loop_t *loop, int can_delay)
+{
+    if (can_delay && loop->_async_io->submission.size < h2o_io_uring_batch_size)
+        return 0;
+
+    int made_progress = 0;
+
+    while (loop->_async_io->submission.head != NULL) {
+        struct io_uring_sqe *sqe;
+        if ((sqe = io_uring_get_sqe(&loop->_async_io->uring)) == NULL)
+            break;
+        struct st_h2o_io_uring_cmd_t *cmd = pop_queue(&loop->_async_io->submission);
+        assert(cmd != NULL);
+        io_uring_prep_splice(sqe, cmd->splice_.filefd, cmd->splice_.offset, cmd->splice_.pipefd, -1, cmd->splice_.len, 0);
+        sqe->user_data = (uint64_t)cmd;
+        made_progress = 1;
+    }
+
+    if (made_progress) {
+        int ret;
+        while ((ret = io_uring_submit(&loop->_async_io->uring)) == -EINTR)
+            ;
+        if (ret < 0)
+            h2o_fatal("io_uring_submit:%s", strerror(-ret));
+    }
+
+    return made_progress;
+}
+
+static int check_completion(h2o_loop_t *loop, struct st_h2o_io_uring_cmd_t *cmd_sync)
+{
+    int cmd_sync_done = 0, ret;
+
+    while (1) {
+        struct st_h2o_io_uring_cmd_t *cmd;
+        int res;
+
+        { /* obtain completed command and its result */
+            struct io_uring_cqe *cqe;
+            while ((ret = io_uring_peek_cqe(&loop->_async_io->uring, &cqe)) == -EINTR)
+                ;
+            if (ret != 0)
+                break;
+            cmd = (struct st_h2o_io_uring_cmd_t *)cqe->user_data;
+            res = cqe->res;
+            io_uring_cqe_seen(&loop->_async_io->uring, cqe);
+        }
+
+        /* Check error. Or if partial read, schedule read of the remainder. */
+        if (res != cmd->splice_.len) {
+            if (res > 0 || res == -EAGAIN) {
+                /* partial read */
+                if (res > 0) {
+                    assert(res < cmd->splice_.len);
+                    cmd->splice_.offset += res;
+                    cmd->splice_.len -= res;
+                }
+                { /* H2O_PULL_SENDVEC_MAX_SIZE is equal to the default pipe buffer size of 64KB, therefore it is likely that we can
+                   * splice entire sendvec at once. However, when the input is not aligned to page boundary, we might need to expand
+                   * the buffer. */
+                    int sz = fcntl(cmd->splice_.pipefd, F_GETPIPE_SZ);
+                    char errbuf[256];
+                    if (sz == -1)
+                        h2o_fatal("fcntl(F_GETPIPE_SZ):%d:%s", errno, h2o_strerror_r(errno, errbuf, sizeof(errbuf)));
+                    sz += 16384;
+                    if (fcntl(cmd->splice_.pipefd, F_SETPIPE_SZ, sz) < sz)
+                        h2o_fatal("fcntl(F_SETPIPE_SZ):%d:%s", errno, h2o_strerror_r(errno, errbuf, sizeof(errbuf)));
```

Comment:
When I do `h2load -m 20 -n 10000 -c 100 -t 6 https://192.168.2.2:8443/tmp/hello1m.txt` I see a crash at this line, complaining that the pipe buffer size cannot be raised to 24576 (EPERM).

Rather than trying to splice the specified amount (and trying to expand the buffer when EAGAIN is returned), it might be better to consider the specified amount as the upper limit and return partial transfers as-is.